proposals/NNNN-support-for-android-custom-routerservice.md

```diff
@@ -0,0 +1,275 @@
+                        config.service = name;
+                    }
+                    transport = new TransportBrokerImpl(contextWeakReference, config.appId, config.service);
+                    // because this callback works asynchrnous, we have to call TransportManager.start here.
+                    transport.start();
+                } else {
+                    enterLegacyMode("Router service is not trusted. Entering legacy mode");
+                }
+            }
+        });
+    }
+
+```
+
+#### Using cache for saving cost to find the connected RouterService.
+When the app finds the connected RouterService, I think it's better to cache the service name on somewhere, e.g. SharedPreferences, so that it won't bother the other app's RouterService too often.
+
+### If an app contains custom RouterService, it must start by himself
+Because custom RouterService cannot rely on other apps to be waken up, it must start RouterService whenever needed:
+The pseudo-code in SdlBroadcastReceiver class looks like this:
+
+```java
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        ...
+        if(action.equalsIgnoreCase(TransportConstants.START_ROUTER_SERVICE_ACTION)){
+            ... // snip
+            // if we need to wake up RotuerService, first check to see if we have the custom RouterService.
+            try {
+                ApplicationInfo appInfo = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
+                if (appInfo.metaData != null) {
+                    boolean isCustom = appInfo.metaData.getBoolean(context.getString(R.string.sdl_router_service_is_custom_name);
+                    if (isCustom) {
+                        // then startService myself.
+                        Intent serviceIntent = new Intent(context, SdlRouterService.class);
+                        context.startService(serviceIntent);
+                    }
+                }
+            } catch(PackageManager.NameNotFoundException e) {}
+        }
+    }
+```
+
+## Potential downsides
+
+- The proposed solution will use SdlRouterStatusProvider, which actually binds to other app's RouterService and asks if the RouterService has connected transports. This will be done in main thread, so the caller should execute it from worker thread. This can be done with AsyncTask anyway.
+- Also when RouterService is bound from SdlRouterStatusProvider, RouterService should not enter to foreground. This is to avoid unneeded notification message coming up especially on Android 8 or above.
+
+## Impact on existing code
+
+- RouterServiceValidator's validate will become asynchronous method.
+- The synchronous validate() method would be deprecated.
+
+## Alternatives considered
+- The solution would be beneficial for both custom RouterService and standard RouterService, because it actually increases the chance to find a RouterService that connects with head unit.
```

Comment:
Update to: 
- The solution would be beneficial for both custom RouterService and standard RouterService, because it actually increases the chance to find a RouterService that connects with **the** head unit.